File: python/src/limiter.py
"""The gate: evaluate provider-style limits before a call, reserve/reconcile
tokens, cap concurrency, and emit standard back-pressure signals.

Design mirrors how model-serving APIs meter in practice:

- Multiple limits per model enforced *together* (RPM + TPM + daily + cost).
- Hierarchical *scopes*: a request must clear the global fleet cap, its tenant's
  cap, and the user's cap - the toughest applicable rule wins.
- Reserve on estimated ``max_tokens``, then commit the actual usage (or refund a
  failed call), so a large requested completion counts against you immediately
  and is reconciled afterwards.
- Concurrency slots cap in-flight requests per scope.
- On denial, a precise ``retry_after`` and an optional cheaper-model fallback, so
  callers can degrade gracefully instead of hard-failing.
"""

from __future__ import annotations

import math
import time
from dataclasses import dataclass, field
from typing import Self

from rules import LimitRule, Scope
from store import COST, REQUESTS, TOKENS, InMemoryStore, Store


@dataclass
class Decision:
    allowed: bool
    model: str
    retry_after: float = 0.0
    tripped_rule: LimitRule | None = None
    scope: Scope | None = None
    remaining: dict[str, float] = field(default_factory=dict)
    suggested_fallback: str | None = None


@dataclass
class Reservation:
    model: str
    tokens: float
    cost: float
    handles: list  # [(key, handle), ...]
    committed: bool = False


@dataclass
class Slot:
    limiter: Limiter
    keys: list
    ok: bool = True
    tripped_rule: LimitRule | None = None

    def release(self) -> None:
        for k in self.keys:
            self.limiter.store.release_concurrency(k)
        self.keys = []

    def __enter__(self) -> Self:
        return self

    def __exit__(self, *exc) -> None:
        self.release()


class Limiter:
    def __init__(self, rules, store: Store | None = None, clock=None, fallbacks=None) -> None:
        self.rules: list[LimitRule] = list(rules)
        self.store: Store = store or InMemoryStore()
        self.clock = clock or time.time
        self.fallbacks: dict[str, str] = dict(fallbacks or {})
        self._by_model: dict[str, list[tuple[int, LimitRule]]] = {}
        for i, r in enumerate(self.rules):
            self._by_model.setdefault(r.model, []).append((i, r))

    def _applicable(self, model: str) -> list[tuple[int, LimitRule]]:
        return self._by_model.get(model, []) + self._by_model.get("*", [])

    def _key(self, index: int, rule: LimitRule, tenant, user):
        if rule.scope is Scope.GLOBAL:
            return (index, "g")
        if rule.scope is Scope.TENANT:
            return (index, "t", tenant)
        return (index, "u", tenant, user)

    def try_acquire(self, model: str, *, tokens: float = 0.0, cost: float = 0.0,
                    tenant=None, user=None, now=None, record: bool = True) -> Decision:
        now = self.clock() if now is None else now
        rules = self._applicable(model)

        worst: tuple[float, LimitRule] | None = None
        for i, r in rules:
            if r.max_tokens is None and r.max_requests is None and r.max_cost is None:
                continue  # e.g. a concurrency-only rule
            key = self._key(i, r, tenant, user)
            used_t, used_r, used_c = self.store.snapshot(key, now, r.window_seconds, r.bucket_seconds)
            breached = []
            if r.max_tokens is not None and used_t + tokens > r.max_tokens + 1e-9:
                breached.append((TOKENS, used_t + tokens - r.max_tokens))
            if r.max_requests is not None and used_r + 1 > r.max_requests + 1e-9:
                breached.append((REQUESTS, used_r + 1 - r.max_requests))
            if r.max_cost is not None and used_c + cost > r.max_cost + 1e-9:
                breached.append((COST, used_c + cost - r.max_cost))
            if breached:
                ra = 0.0
                for dim, over in breached:
                    ra = max(ra, self.store.time_to_free(
                        key, now, r.window_seconds, r.bucket_seconds, over, dim))
                if worst is None or ra > worst[0]:
                    worst = (ra, r)

        if worst is not None:
            ra, rule = worst
            return Decision(False, model, retry_after=ra, tripped_rule=rule, scope=rule.scope,
                            suggested_fallback=self.fallbacks.get(model))

        if record:
            for i, r in rules:
                if r.max_tokens is None and r.max_requests is None and r.max_cost is None:
                    continue
                key = self._key(i, r, tenant, user)
                self.store.add(key, now, r.window_seconds, r.bucket_seconds, tokens, 1.0, cost)

        return Decision(True, model, remaining=self._remaining(model, tenant, user, now))

    def _remaining(self, model, tenant, user, now) -> dict[str, float]:
        rem: dict[str, float] = {}

        def tighten(name, value):
            rem[name] = min(rem.get(name, float("inf")), value)

        for i, r in self._applicable(model):
            key = self._key(i, r, tenant, user)
            used_t, used_r, used_c = self.store.snapshot(key, now, r.window_seconds, r.bucket_seconds)
            if r.max_tokens is not None:
                tighten("tokens", r.max_tokens - used_t)
            if r.max_requests is not None:
                tighten("requests", r.max_requests - used_r)
            if r.max_cost is not None:
                tighten("cost", r.max_cost - used_c)
        return {k: max(0.0, v) for k, v in rem.items()}

    # ---- reserve -> commit / refund ----

    def reserve(self, model: str, *, tokens: float = 0.0, cost: float = 0.0,
                tenant=None, user=None, now=None) -> tuple[Decision, Reservation | None]:
        now = self.clock() if now is None else now
        decision = self.try_acquire(model, tokens=tokens, cost=cost, tenant=tenant,
                                    user=user, now=now, record=False)
        if not decision.allowed:
            return decision, None
        handles = []
        for i, r in self._applicable(model):
            if r.max_tokens is None and r.max_requests is None and r.max_cost is None:
                continue
            key = self._key(i, r, tenant, user)
            h = self.store.add(key, now, r.window_seconds, r.bucket_seconds, tokens, 1.0, cost)
            handles.append((key, h))
        return decision, Reservation(model=model, tokens=tokens, cost=cost, handles=handles)

    def commit(self, reservation: Reservation, *, actual_tokens=None, actual_cost=None) -> None:
        if reservation.committed:
            return
        dt = 0.0 if actual_tokens is None else actual_tokens - reservation.tokens
        dc = 0.0 if actual_cost is None else actual_cost - reservation.cost
        for key, handle in reservation.handles:
            self.store.adjust(key, handle, dt, 0.0, dc)
        reservation.tokens += dt
        reservation.cost += dc
        reservation.committed = True

    def refund(self, reservation: Reservation) -> None:
        if reservation.committed:
            return
        for key, handle in reservation.handles:
            self.store.adjust(key, handle, -reservation.tokens, -1.0, -reservation.cost)
        reservation.handles = []
        reservation.committed = True

    # ---- concurrency slots ----

    def acquire_slot(self, model: str, *, tenant=None, user=None) -> Slot:
        acquired = []
        for i, r in self._applicable(model):
            if r.max_concurrent is None:
                continue
            key = ("conc",) + self._key(i, r, tenant, user)
            if not self.store.try_add_concurrency(key, r.max_concurrent):
                for k in acquired:
                    self.store.release_concurrency(k)
                return Slot(self, [], ok=False, tripped_rule=r)
            acquired.append(key)
        return Slot(self, acquired, ok=True)

    # ---- graceful degradation ----

    def acquire_or_fallback(self, model: str, **kwargs) -> tuple[Decision, str]:
        decision = self.try_acquire(model, **kwargs)
        if decision.allowed:
            return decision, model
        fallback = self.fallbacks.get(model)
        if fallback:
            alt = self.try_acquire(fallback, **kwargs)
            if alt.allowed:
                return alt, fallback
        return decision, model


def rate_limit_headers(decision: Decision, rule: LimitRule | None = None) -> dict[str, str]:
    """Render a decision as the ``X-RateLimit-*`` / ``Retry-After`` headers real
    providers return, so you can pass back-pressure straight to your caller."""
    rule = rule or decision.tripped_rule
    headers: dict[str, str] = {}
    if decision.retry_after > 0:
        headers["Retry-After"] = str(math.ceil(decision.retry_after))
    if rule is not None:
        if rule.max_requests is not None:
            headers["X-RateLimit-Limit-Requests"] = str(int(rule.max_requests))
        if rule.max_tokens is not None:
            headers["X-RateLimit-Limit-Tokens"] = str(int(rule.max_tokens))
    for name, value in (decision.remaining or {}).items():
        headers[f"X-RateLimit-Remaining-{name.capitalize()}"] = str(int(value))
    return headers

File: ts/src/limiter.ts
/**
 * The gate: evaluate provider-style limits before a call, reserve/reconcile
 * tokens, cap concurrency, and emit standard back-pressure signals.
 *
 * Design mirrors how model-serving APIs meter in practice:
 *
 * - Multiple limits per model enforced *together* (RPM + TPM + daily + cost).
 * - Hierarchical *scopes*: a request must clear the global fleet cap, its tenant's
 *   cap, and the user's cap - the toughest applicable rule wins.
 * - Reserve on estimated `maxTokens`, then commit the actual usage (or refund a
 *   failed call), so a large requested completion counts against you immediately
 *   and is reconciled afterwards.
 * - Concurrency slots cap in-flight requests per scope.
 * - On denial, a precise `retryAfter` and an optional cheaper-model fallback, so
 *   callers can degrade gracefully instead of hard-failing.
 */

import { LimitRule, Scope } from "./rules";
import { COST, Handle, InMemoryStore, REQUESTS, Store, TOKENS } from "./store";

/** The result of an admission check. */
export interface Decision {
  allowed: boolean;
  model: string;
  retryAfter: number;
  trippedRule: LimitRule | null;
  scope: Scope | null;
  remaining: Record<string, number>;
  suggestedFallback: string | null;
}

/**
 * Records counters up front so the actual usage can be reconciled (commit) or
 * released (refund) afterwards.
 */
export interface Reservation {
  model: string;
  tokens: number;
  cost: number;
  committed: boolean;
  handles: Array<[string, Handle]>;
}

/** The optional arguments of an admission call. */
export interface AcquireOptions {
  tokens?: number;
  cost?: number;
  tenant?: string;
  user?: string;
  now?: number;
}

/** The actual usage passed to {@link Limiter.commit}. */
export interface CommitOptions {
  actualTokens?: number;
  actualCost?: number;
}

/** Options for constructing a {@link Limiter}. */
export interface LimiterOptions {
  store?: Store;
  clock?: () => number;
  fallbacks?: Record<string, string>;
}

/**
 * A held concurrency reservation. Call {@link Slot.release} (e.g. in a `finally`)
 * to give the in-flight slots back.
 */
export class Slot {
  ok: boolean;
  trippedRule: LimitRule | null;
  private readonly limiter: Limiter;
  private keys: string[];

  constructor(limiter: Limiter, keys: string[], ok: boolean, trippedRule: LimitRule | null = null) {
    this.limiter = limiter;
    this.keys = keys;
    this.ok = ok;
    this.trippedRule = trippedRule;
  }

  /** Returns every concurrency slot this Slot holds. */
  release(): void {
    for (const k of this.keys) {
      this.limiter.store.releaseConcurrency(k);
    }
    this.keys = [];
  }
}

/**
 * The gate: it evaluates limits before a call, reserves/reconciles tokens, caps
 * concurrency, and emits standard back-pressure signals.
 */
export class Limiter {
  readonly rules: LimitRule[];
  readonly store: Store;
  private readonly clock: () => number;
  private readonly fallbacks: Record<string, string>;
  private readonly byModel = new Map<string, Array<[number, LimitRule]>>();

  constructor(rules: LimitRule[], opts: LimiterOptions = {}) {
    this.rules = [...rules];
    this.store = opts.store ?? new InMemoryStore();
    this.clock = opts.clock ?? (() => Date.now() / 1000);
    this.fallbacks = { ...(opts.fallbacks ?? {}) };
    this.rules.forEach((r, i) => {
      const list = this.byModel.get(r.model) ?? [];
      list.push([i, r]);
      this.byModel.set(r.model, list);
    });
  }

  private applicable(model: string): Array<[number, LimitRule]> {
    return [...(this.byModel.get(model) ?? []), ...(this.byModel.get("*") ?? [])];
  }

  private key(index: number, rule: LimitRule, tenant: string, user: string): string {
    if (rule.scope === Scope.Global) {
      return `${index}|g`;
    }
    if (rule.scope === Scope.Tenant) {
      return `${index}|t|${tenant}`;
    }
    return `${index}|u|${tenant}|${user}`;
  }

  private resolveNow(now: number | undefined): number {
    return now === undefined ? this.clock() : now;
  }

  /**
   * Evaluates a call against every applicable rule and, unless denied, records
   * it. On denial the Decision carries the retryAfter, the tripped rule, its
   * scope, and any configured fallback.
   */
  tryAcquire(model: string, opts: AcquireOptions = {}): Decision {
    return this.tryAcquireInner(model, opts, true);
  }

  private tryAcquireInner(model: string, opts: AcquireOptions, record: boolean): Decision {
    const now = this.resolveNow(opts.now);
    const tokens = opts.tokens ?? 0;
    const cost = opts.cost ?? 0;
    const tenant = opts.tenant ?? "";
    const user = opts.user ?? "";
    const rules = this.applicable(model);

    let worstRetry = -1;
    let worstRule: LimitRule | null = null;
    for (const [i, r] of rules) {
      if (!r.hasUsageLimit()) {
        continue; // e.g. a concurrency-only rule
      }
      const key = this.key(i, r, tenant, user);
      const [usedT, usedR, usedC] = this.store.snapshot(key, now, r.windowSeconds, r.bucketSeconds);
      const breached: Array<[number, number]> = [];
      if (r.maxTokens !== undefined && usedT + tokens > r.maxTokens + 1e-9) {
        breached.push([TOKENS, usedT + tokens - r.maxTokens]);
      }
      if (r.maxRequests !== undefined && usedR + 1 > r.maxRequests + 1e-9) {
        breached.push([REQUESTS, usedR + 1 - r.maxRequests]);
      }
      if (r.maxCost !== undefined && usedC + cost > r.maxCost + 1e-9) {
        breached.push([COST, usedC + cost - r.maxCost]);
      }
      if (breached.length > 0) {
        let ra = 0;
        for (const [dim, over] of breached) {
          ra = Math.max(ra, this.store.timeToFree(key, now, r.windowSeconds, r.bucketSeconds, over, dim));
        }
        if (ra > worstRetry) {
          worstRetry = ra;
          worstRule = r;
        }
      }
    }

    if (worstRule !== null) {
      return {
        allowed: false,
        model,
        retryAfter: worstRetry,
        trippedRule: worstRule,
        scope: worstRule.scope,
        remaining: {},
        suggestedFallback: this.fallbacks[model] ?? null,
      };
    }

    if (record) {
      for (const [i, r] of rules) {
        if (!r.hasUsageLimit()) {
          continue;
        }
        const key = this.key(i, r, tenant, user);
        this.store.add(key, now, r.windowSeconds, r.bucketSeconds, tokens, 1.0, cost);
      }
    }

    return {
      allowed: true,
      model,
      retryAfter: 0,
      trippedRule: null,
      scope: null,
      remaining: this.remaining(model, tenant, user, now),
      suggestedFallback: null,
    };
  }

  private remaining(model: string, tenant: string, user: string, now: number): Record<string, number> {
    const rem: Record<string, number> = {};
    const tighten = (name: string, value: number): void => {
      rem[name] = Math.min(rem[name] ?? Infinity, value);
    };
    for (const [i, r] of this.applicable(model)) {
      const key = this.key(i, r, tenant, user);
      const [usedT, usedR, usedC] = this.store.snapshot(key, now, r.windowSeconds, r.bucketSeconds);
      if (r.maxTokens !== undefined) {
        tighten("tokens", r.maxTokens - usedT);
      }
      if (r.maxRequests !== undefined) {
        tighten("requests", r.maxRequests - usedR);
      }
      if (r.maxCost !== undefined) {
        tighten("cost", r.maxCost - usedC);
      }
    }
    for (const k of Object.keys(rem)) {
      rem[k] = Math.max(0, rem[k]);
    }
    return rem;
  }

  // ---- reserve -> commit / refund ----

  /**
   * Records the estimate up front and returns a Reservation to reconcile later.
   * If the call would be denied, it returns the denied Decision and a null
   * Reservation.
   */
  reserve(model: string, opts: AcquireOptions = {}): [Decision, Reservation | null] {
    const now = this.resolveNow(opts.now);
    const decision = this.tryAcquireInner(model, { ...opts, now }, false);
    if (!decision.allowed) {
      return [decision, null];
    }
    const tokens = opts.tokens ?? 0;
    const cost = opts.cost ?? 0;
    const tenant = opts.tenant ?? "";
    const user = opts.user ?? "";
    const handles: Array<[string, Handle]> = [];
    for (const [i, r] of this.applicable(model)) {
      if (!r.hasUsageLimit()) {
        continue;
      }
      const key = this.key(i, r, tenant, user);
      const h = this.store.add(key, now, r.windowSeconds, r.bucketSeconds, tokens, 1.0, cost);
      handles.push([key, h]);
    }
    return [decision, { model, tokens, cost, committed: false, handles }];
  }

  /**
   * Reconciles a reservation to the actual usage. An undefined actualTokens or
   * actualCost leaves that dimension at the reserved estimate. A committed
   * reservation is frozen and further commits/refunds are no-ops.
   */
  commit(res: Reservation, opts: CommitOptions = {}): void {
    if (res.committed) {
      return;
    }
    const dt = opts.actualTokens === undefined ? 0 : opts.actualTokens - res.tokens;
    const dc = opts.actualCost === undefined ? 0 : opts.actualCost - res.cost;
    for (const [key, handle] of res.handles) {
      this.store.adjust(key, handle, dt, 0.0, dc);
    }
    res.tokens += dt;
    res.cost += dc;
    res.committed = true;
  }

  /** Releases everything a reservation recorded, e.g. after a failed call. */
  refund(res: Reservation): void {
    if (res.committed) {
      return;
    }
    for (const [key, handle] of res.handles) {
      this.store.adjust(key, handle, -res.tokens, -1.0, -res.cost);
    }
    res.handles = [];
    res.committed = true;
  }

  // ---- concurrency slots ----

  /**
   * Reserves one concurrency slot per applicable max-concurrent rule. If any rule
   * is full, all partially-acquired slots are released and the returned Slot has
   * ok=false with the tripped rule.
   */
  acquireSlot(model: string, opts: AcquireOptions = {}): Slot {
    const tenant = opts.tenant ?? "";
    const user = opts.user ?? "";
    const acquired: string[] = [];
    for (const [i, r] of this.applicable(model)) {
      if (r.maxConcurrent === undefined) {
        continue;
      }
      const key = "conc|" + this.key(i, r, tenant, user);
      if (!this.store.tryAddConcurrency(key, r.maxConcurrent)) {
        for (const k of acquired) {
          this.store.releaseConcurrency(k);
        }
        return new Slot(this, [], false, r);
      }
      acquired.push(key);
    }
    return new Slot(this, acquired, true, null);
  }

  // ---- graceful degradation ----

  /**
   * Tries the model, then its configured fallback on denial, and reports which
   * model was ultimately chosen.
   */
  acquireOrFallback(model: string, opts: AcquireOptions = {}): [Decision, string] {
    const decision = this.tryAcquire(model, opts);
    if (decision.allowed) {
      return [decision, model];
    }
    const fallback = this.fallbacks[model];
    if (fallback !== undefined) {
      const alt = this.tryAcquire(fallback, opts);
      if (alt.allowed) {
        return [alt, fallback];
      }
    }
    return [decision, model];
  }
}

/**
 * Renders a decision as the `X-RateLimit-*` / `Retry-After` headers real
 * providers return, so you can pass back-pressure straight to your caller.
 */
export function rateLimitHeaders(decision: Decision, rule: LimitRule | null = null): Record<string, string> {
  const r = rule ?? decision.trippedRule;
  const headers: Record<string, string> = {};
  if (decision.retryAfter > 0) {
    headers["Retry-After"] = String(Math.ceil(decision.retryAfter));
  }
  if (r !== null) {
    if (r.maxRequests !== undefined) {
      headers["X-RateLimit-Limit-Requests"] = String(Math.trunc(r.maxRequests));
    }
    if (r.maxTokens !== undefined) {
      headers["X-RateLimit-Limit-Tokens"] = String(Math.trunc(r.maxTokens));
    }
  }
  for (const [name, value] of Object.entries(decision.remaining)) {
    const capitalized = name.charAt(0).toUpperCase() + name.slice(1);
    headers[`X-RateLimit-Remaining-${capitalized}`] = String(Math.trunc(value));
  }
  return headers;
}
